ecds: hook UDP session filters with config provider

ohadvano opened the pull request:
Additional Description: Adding a new ``ConfigProviderManager`` for UDP session filters with tests. Consuming this provider for static filters. This PR does not add dynamic filters yet, will be done in next PR.
Risk Level: low
Testing: unit tests
Docs Changes: none
Release Notes: none
Platform Specific Features: none

ohadvano commented:
@RyanTheOptimist could you please take this one, as you review the previous?

adisuissa commented:
Assigning Ryan, but please let me know if there's someone else that you wish to review this.
/assign @RyanTheOptimist

RyanTheOptimist commented:
/assign @htuch 
Harvey, you're a lot more familiar with xDS than I am. Can you review this?

adisuissa commented:
> /assign @htuch Harvey, you're a lot more familiar with xDS than I am. Can you review this?

I should probably review it, and pass it to Kuat later.
Unassigning Harvey for now.

adisuissa reviewed (approved):
LGTM, thanks!

/assign-from @envoyproxy/senior-maintainers

adisuissa commented:
Also assigning Kuat who has lots of experience.
/assign @kyessenov

kyessenov commented on source/extensions/filters/udp/udp_proxy/config.h:
```diff
@@ -151,8 +158,12 @@ class UdpProxyFilterConfigImpl : public UdpProxyFilterConfig,
-      factory(callbacks);
+    for (const auto& filter_config_provider : filter_factories_) {
+      auto config = filter_config_provider->config();
+      if (config.has_value()) {
```
Should there be some warning/error for a missing config here?

ohadvano replied:
agree, added a log

kyessenov replied:
Thanks. Can you update the docs in the UDP listener, saying that the missing config in ECDS results in a no-op warning? For other filters, it is protocol-specific, e.g. 503 for HTTP.

ohadvano replied:
ECDS isn't really getting supported in this PR - the API remains the same, requiring consumers to configure the filter statically. In next PR I'll add dynamic consumption that uses discovery, and I'll update docs accordingly.
Also note that this kind of change applies to UDP **session** filters and not regular UDP listener filters

adisuissa commented:
Seems that coverage needs to be improved.
See https://storage.googleapis.com/envoy-pr/02302c0/coverage/source/extensions/filters/udp/udp_proxy/index.html

Also adding @htuch for senior-maintainer stamp instead of @wbpcode who is OOO.
/assign @htuch

ohadvano commented:
@kyessenov I removed the log line I added for two reasons:
1. Coverage is reduced below minimum, and I'm not sure if this scenario is reproducible in test (see 2nd reason).
2. Not sure if we would ever get there. For static filters - the filter factory validates the protobuf message, so if the config is missing, an exception will be thrown at bootstrap. On dynamic updates (not functional yet), xDS will also throw an exception if config update is invalid (factory will also try to instantiate the filter and fail)

ohadvano commented:
@adisuissa did you mean to assign @htuch?
No senior maintainer is currently assigned

adisuissa commented:
> @adisuissa did you mean to assign @htuch? No senior maintainer is currently assigned

Nope.. race between repokitteh and github's UI. sorry.

kyessenov commented:
> @kyessenov I removed the log line I added for two reasons:
> 
> 1. Coverage is reduced below minimum, and I'm not sure if this scenario is reproducible in test (see 2nd reason).
> 2. Not sure if we would ever get there. For static filters - the filter factory validates the protobuf message, so if the config is missing, an exception will be thrown at bootstrap. On dynamic updates (not functional yet), xDS will also throw an exception if config update is invalid (factory will also try to instantiate the filter and fail)

This path used when ECDS fails to push config and the parent LDS was not configured to warm on ECDS. I think we have to handle this case, but I'm OK waiting till the final integration.

ohadvano commented:
I see, makes sense. I'll add it back once in the dynamic configs PR with test that would be able to cover it using ECDS
